Procedural Map Generator Improvements [Perfect for Merging/Porting]

RemieRichards opened the pull request:
- Fixes a bug where turf cluster checks checked for atoms instead of turfs
- Changes to the clusterCheckFlag system by replacing + adding new defines
- Adds a proc `defineCircularRegion()` to define a circular map region, CIRCLES and SPHERES!
- `defineRegion()` and `defineCircularRegion()` no longer replace the map eniterely, they now add all the new turfs to the map list (no duplicates), the ability to overwrite the map has been moved to a new argument on both procs `replace`
- Adds a proc `undefineRegion()` which just makes the map list a new list, it's just there to look nice
- Removes bottomLeft and topRight, they weren't even needed
- Made the turf clustering checks faster by remembering that `trange()` exists, and actually using it.

To all the codebases that took the map generator last time, feel free to take the improvements too.
I know VG are doing an entire procedural map using this, so you'll definitely want to pick it up.

RemieRichards committed:
Fixes a bug where turf cluster checks checked for atoms instead of turfs, adds proc to define a circular region, CIRCLES!

RemieRichards committed:
Made defineRegion() procs additive instead of overwriting, overwriting has been moved to a "replace" argument, adds undefineRegion() which makes the map list a new list.

RemieRichards commented:
Ok major bugs at the moment, Already working on fixes.

tigercat2000 commented:
I look forward to it.

astralenigma commented on code/modules/procedural mapping/mapGeneratorModule.dm:
```diff
@@ -38,24 +49,47 @@
+					if(locate(atomPath) in range(clustering,T))
 						continue
-			if(prob(spawnableAtoms[atomPath]))
-				new atomPath (T)
```
What did those 2 lines do before deleting?

RemieRichards replied:
Thanks for this, massive derp on my part.

RemieRichards committed:
Fixes atoms not spawning, Corrects defineCircularRange() to produce proper spheres when given differing Z coords.

RemieRichards committed:
Adjusts some code to reduce the rapid increase of clustering accidentally somehow caused in this improvement PR

RemieRichards commented:
Alright fixed the lack of atoms, it was a stupid issue.
Corrected defineCircularRegion() to function correctly when given 3D coords (Proper spheres)

astralenigma commented on code/modules/procedural mapping/mapGeneratorModule.dm:
```diff
@@ -38,24 +34,61 @@
+
+		//Success!
 		if(prob(spawnableTurfs[turfPath]))
 			T.ChangeTurf(turfPath)
```
Shouldn't this if only be checked after the other checks?

RemieRichards replied:
This isn't an issue, this is never reached if the previous checks fail, due to intelligent use of continue and break statements.

RemieRichards committed:
Fixes the scope of the 4 Continue statements to correctly end the turfPath and atomPath loops, Corrects define bitflags

RemieRichards commented:
## All improvements added, Producing results identically to before
## I am happy with it once more!

GunHog commented:
So, what do we...do with this?

RemieRichards commented:
@GunHog well, nothing really, /tg/ hasn't done anything with the generator system, but I used it on the servers to produce forests of about 200x200 size in 2-3 seconds.

/vg/ are using it for an entirely custom map

This is just some improvements/fixes to the system.

tigercat2000 commented:
@RemieRichards By the way, not sure if a result of the port- selecting none on the nature debug cluster menu makes it return 'invalid'

HoaxIT commented:
i love this, porting to hippie.

RemieRichards commented:
@tigercat2000 Oh yes, That was a minor fix I never got round to fixing.
I'm not sure why you'd ever want to use no clustering on the nature generator, but then again I don't know why you'd want 3D sphere maps yet I added support for that :P

RemieRichards committed:
Fixes the nature generator verb failing to recognise CLUSTER_CHECK_NONE as a valid flag

RemieRichards commented:
@HoaxIT Enjoy it :)

Codec-pedras commented:
How easy it is to make code mechanical automatic developer ? with many of the walls and parts ?

Dorsisdwarf commented:
What

Miauw62 commented:
how is code formed? how Z get inetializid?

RemieRichards commented:
@Miauw62 When 2 operands love each other very much, they get an operator and they position themselves either side of it.

that is how code formed.

PJB3005 commented:
Can I port this to /vg/ again?

RemieRichards commented:
@PJB3005 

>To all the codebases that took the map generator last time, feel free to take the improvements too.
I know VG are doing an entire procedural map using this, so you'll definitely want to pick it up.

PJB3005 commented:
oh, oops, anyways thanks!

RemieRichards committed:
Fixes centerX and centerY not being correctly centered, corrects sphere logic.

why is it called sphereMagic? because as far as I'm concerned, it is magic, it was previously named centerZ but it actually was quite a ways off being the center of any spheres. Idk what it does, but It serves a vital part of building mathematically correct spheres.

RemieRichards committed:
minor correction.
